pandas/core/strings/accessor.py

```diff
@@ -0,0 +1,3078 @@
+    happen if the arrays in list_of columns have the wrong dtypes or content.
+
+    Parameters
+    ----------
+    list_of_columns : list of numpy arrays
+        List of arrays to be concatenated with sep;
+        these arrays may not contain NaNs!
+    sep : string
+        The separator string for concatenating the columns.
+
+    Returns
+    -------
+    nd.array
+        The concatenation of list_of_columns with sep.
+    """
+    try:
+        result = cat_core(list_of_columns, sep)
+    except TypeError:
+        # if there are any non-string values (wrong dtype or hidden behind
+        # object dtype), np.sum will fail; catch and return with better message
+        for column in list_of_columns:
+            dtype = lib.infer_dtype(column, skipna=True)
+            if dtype not in ["string", "empty"]:
+                raise TypeError(
+                    "Concatenation requires list-likes containing only "
+                    "strings (or missing values). Offending values found in "
+                    f"column {dtype}"
+                ) from None
+    return result
+
+
+def cat_core(list_of_columns: List, sep: str):
+    """
+    Auxiliary function for :meth:`str.cat`
+
+    Parameters
+    ----------
+    list_of_columns : list of numpy arrays
+        List of arrays to be concatenated with sep;
+        these arrays may not contain NaNs!
+    sep : string
+        The separator string for concatenating the columns.
+
+    Returns
+    -------
+    nd.array
+        The concatenation of list_of_columns with sep.
+    """
+    if sep == "":
+        # no need to interleave sep if it is empty
+        arr_of_cols = np.asarray(list_of_columns, dtype=object)
+        return np.sum(arr_of_cols, axis=0)
+    list_with_sep = [sep] * (2 * len(list_of_columns) - 1)
+    list_with_sep[::2] = list_of_columns
+    arr_with_sep = np.asarray(list_with_sep, dtype=object)
+    return np.sum(arr_with_sep, axis=0)
+
+
+def _groups_or_na_fun(regex):
+    """Used in both extract_noexpand and extract_frame"""
+    if regex.groups == 0:
+        raise ValueError("pattern contains no capture groups")
+    empty_row = [np.nan] * regex.groups
+
+    def f(x):
+        if not isinstance(x, str):
+            return empty_row
+        m = regex.search(x)
+        if m:
+            return [np.nan if item is None else item for item in m.groups()]
+        else:
+            return empty_row
+
+    return f
+
+
+def _result_dtype(arr):
+    # workaround #27953
+    # ideally we just pass `dtype=arr.dtype` unconditionally, but this fails
+    # when the list of values is empty.
+    from pandas.core.arrays.string_ import StringDtype
+
+    if isinstance(arr.dtype, StringDtype):
+        return arr.dtype.name
+    else:
+        return object
+
+
+def _get_single_group_name(rx):
+    try:
+        return list(rx.groupindex.keys()).pop()
+    except IndexError:
+        return None
+
+
+def _str_extract_noexpand(arr, pat, flags=0):
```

Comment:
See @TomAugspurger note in the top post: 

> Note that there are a few methods like cat, extract, and extractall that don't yet dispatch. I need to look into them a bit more, there implementation is a bit confusing.

Reply:
ok i understand and that is fine. but for the purposes of this PR it is very confusing to put essentially an implementation in the accessor.py, rather than in the concrete classes themselves. alt put them in a separete module for now.

Reply:
> it is very confusing to put essentially an implementation in the accessor.py

I don't think this is confusing, as `extract` is a method of the accessor defined above in this file .. Note that before this PR *all* implementations were in the file where the accessor was defined.

Reply:
Right.

The issue with `extract` specifically is that it operates on the Series rather than the Array. The dispatched functions are all dispatching to the array. I'm sure that we can eventually dispatch to the array by extracting the relevant pieces, but that would be a bit larger change. I've tried to keep the changes here as small as possible (despite the size of the diff) by just moving things and adding just the necessary new pieces.

Reply:
ok fair enough, can you create an issue to separate these other methods (maybe with check boxes) & align witht he strutcure)